File: day05.py
def fully_react_polymer(polymer):
    polymer = list(polymer)
    start_index = 0
    while True:
        for index in range(start_index, len(polymer)):
            unit1 = polymer[index]
            try:
                unit2 = polymer[index + 1]
            except IndexError:
                # reached the end
                pass
            if unit1.lower() == unit2.lower() and unit1 != unit2:
                del polymer[index:index + 2]
                # Restart from just before this reaction.
                start_index = 0 if index == 0 else index - 1
                break
        else:
            return polymer


def shortest_length_by_removing(polymer):
    # pypy 5.10 doesn't have math.inf
    shortest = float('inf')
    for unit in set(polymer.lower()):
        removed_polymer = polymer.replace(unit.upper(), '').replace(unit.lower(), '')
        reacted_polymer_length = len(fully_react_polymer(removed_polymer))
        shortest = reacted_polymer_length if reacted_polymer_length < shortest else shortest
    return shortest



if __name__ == '__main__':
    with open('day05-input.txt', 'r') as input_file:
        polymer = input_file.readline().strip()
        reacted_polymer = fully_react_polymer(polymer)
        # pt 1
        print(len(reacted_polymer))
        # pt 2
        print(shortest_length_by_removing(polymer))

File: day05.js
const fs = require('fs')


function fullyReactPolymer(polymer) {
  const reactedPolymer = polymer.split('')
  let startIndex = 0
  while (true) {
    let done = true
    for (let i = startIndex; i < reactedPolymer.length; i++) {
      const unit1 = reactedPolymer[i]
      const unit2 = reactedPolymer[i + 1]
      if (unit2 === undefined) {
        continue
      }
      if (unit1.toLowerCase() == unit2.toLowerCase() && unit1 !== unit2) {
        reactedPolymer.splice(i, 2)
        startIndex = i === 0 ? 0 : i - 1
        done = false
        break
      }
    }
    if (done) return reactedPolymer
  }
}

function shortestPolymerLength(polymer) {
  const unitsToReplace = new Set(polymer.toLowerCase())
  const shortest = [...unitsToReplace]
    .reduce((acc, unit) => {
      const re = new RegExp(`(${unit.toLowerCase()}|${unit.toUpperCase()})`, 'g')
      const newPolymer = polymer.replace(re, '')
      const length = fullyReactPolymer(newPolymer).length
      return length < acc ? length : acc
  }, Infinity)
  return shortest
}

const polymer = fs.readFileSync('day05-input.txt', 'utf8')
  .trim()
// pt 1
console.log(fullyReactPolymer(polymer).length)
// pt 2
console.log(shortestPolymerLength(polymer))
